presto-hive/src/main/java/com/facebook/presto/hive/CachingDirectoryLister.java

```diff
@@ -98,38 +101,12 @@ private static SchemaTableName parseTableName(String tableName)
         }
 
         if (!cachedTableNames.contains(schemaTableName)) {
-            HadoopFileInfoIterator fileIterator = new HadoopFileInfoIterator(iterator);
-            return new HiveFileIterator(path, p -> fileIterator, namenodeStats, nestedDirectoryPolicy);
+            return delegate.list(fileSystem, table, path, namenodeStats, nestedDirectoryPolicy);
         }
         HadoopFileInfoIterator fileIterator = new HadoopFileInfoIterator(cachingRemoteIterator(iterator, path));
```

Comment:
Is it possible to cache `HiveFileInfo`, for example `Cache<Path, List<HiveFileInfo>> cache`? This way we don't need the wrappers around iterator, and thus could directly use the result from `delegete#list`.

Reply:
yep, cache `HiveFileInfo` could simplify logic for cache hit. I think we still need `cachingRemoteIterator`, which is a clean way for cache miss and backfill the cache.